### PR TITLE
Switch to using spans when doing SourceText.ContentEquals

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Runtime already optimizes these span sequence equal checks.  This will become vectorized calls that end up calling into: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs,998a36a55f580ab1

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@333fred @jjonescz ptal.  Thanks!

File: src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs [hunk 1]
```diff
@@ -212,6 +212,81 @@ public void ContentEquals()
...
+            var sourceText2 = new StringText("?", Encoding.ASCII); // what a non-ascii char will map to.
+
+            var checksum1 = sourceText1.GetChecksum();
+            var checksum2 = sourceText2.GetChecksum();
+
+            Assert.True(checksum1.SequenceEqual(checksum2));
+
+            // This should return false, but is thrown off by the calls to GetChecksum above.  This is a bad bug that
+            // will be fixed once we resolve https://github.com/dotnet/roslyn/issues/70752
+            Assert.False(sourceText1.ContentEquals(sourceText2));
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs:287]: This assert currently fails.  Which is BAD.  This test shows how though even though we have entirely different contents, the improper use of checksums makes us think the contents are actually the same.

File: src/Compilers/Core/Portable/Text/SourceText.cs [hunk 2]
```diff
@@ -1067,24 +1067,19 @@ protected virtual bool ContentEqualsImpl(SourceText other)
 
             var buffer1 = s_charArrayPool.Allocate();
             var buffer2 = s_charArrayPool.Allocate();
+            Debug.Assert(buffer1.Length == buffer2.Length);
+
+            var bufferLength = buffer1.Length;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Text/SourceText.cs:1072]: >var bufferLength = buffer1.Length [](http://example.com/codeflow?start=12&length=33)

Can we use ```CharBufferSize``` constant instead?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

File: src/Compilers/Core/Portable/Text/SourceText.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Text/SourceText.cs]: sgtm :)

**Review by jaredpar — APPROVED** [review]

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

File: src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs [hunk 3]
```diff
@@ -212,6 +212,81 @@ public void ContentEquals()
...
+                    if (encoding1 == Encoding.ASCII && encoding2 == Encoding.ASCII)
+                        Assert.True(sourceText1.GetChecksum().SequenceEqual(sourceText2.GetChecksum()));
+                }
+            }
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/70752")]
+        public void ContentEqualUnaffectedByEncoding2()
+        {
+            var sourceText1 = new StringText("；", Encoding.ASCII); // chinese semicolon
```
Comment (333fred) [inline src/Compilers/Core/CodeAnalysisTest/Text/SourceTextTests.cs:277]: image
Thanks VSCode

**Review by 333fred — APPROVED** [review]

**Pull request merged**